Incorporate Project Space Dialer through AWS CCP package

ctsims opened the pull request:
![image](https://user-images.githubusercontent.com/155066/87215234-7fd15e80-c302-11ea-9a89-fde0d6151e92.png)

##### SUMMARY
Provides an AWS Connect CCP dialer inside of domains when the flag is enabled and configured.

The dialer can be triggered from within a form. For domains where the dialer is enabled, Markdown text in formplayer which produces a `tel:XXXXX` hyperlink is intercepted before rendering and replaced with a reference which will call the domain specific dialer if it is available. This hyperlink will spawn a new tab if no other dialer has been spawned, or will reuse the existing tab if a dialer was spawned previously from web apps.

![image](https://user-images.githubusercontent.com/155066/87215272-e6567c80-c302-11ea-88f6-63b51696d81f.png)

![image](https://user-images.githubusercontent.com/155066/87215276-ece4f400-c302-11ea-99b6-5ea3dcb6d766.png)

The dialer can be configured per project space with an AWS Connect instance, and header text. The Dialer page icon reuses the Custom Project Space Icon if it is available. 

![image](https://user-images.githubusercontent.com/155066/87215292-464d2300-c303-11ea-953b-65cd8b34b201.png)

NOTE - The code for the dialer (static HTML / CSS and JavaScript) is boilerplate from AWS. This is somewhat custom widget for now and I don't think it's worth rebuilding them for now v. building a cleaner experience.

##### FEATURE FLAG
Yes, this creates a new feature flag for the dialer

##### RISK ASSESSMENT / QA PLAN
We should likely run QA on formplayer in general, but the project team can take responsibility for testing the dialer functionality since it is a private, flagged behavior.

ctsims committed:
First checkin for basic models and app

ctsims committed:
Improved templating for widgets

ctsims committed:
Merge branch 'master' into ctsims/call_center_widgets

ctsims committed:
End-to-end functional

ctsims committed:
cleanup/linting

ctsims committed:
linting

ctsims committed:
generify boilerplate

ctsims committed:
Replace logo with custom HQ domain logo

ctsims committed:
generification

ctsims committed:
linting

ctsims committed:
Require toggle for dialer

ctsims committed:
linting

ctsims committed:
remove redundant utils

ctsims commented on corehq/apps/widget/util.py:
```diff
@@ -0,0 +1,11 @@
+from corehq.apps.widget.models import DialerSettings
+
+def domain_uses_dialer(domain):
+    try:
```
@orangejenny I'd like to make it possible for "enterprise spaces to check a box that says "Use Enterprise Dialer settings" which this method would then use to look up the chain at the "control" space to get its settings and mirror them back down. Is that consistent with how enterprise managed domains work?

orangejenny replied:
Yes and no. You can look up the controlling domain's name with `DomainPermissionsMirror.source_domain(domain)` and then pull its settings, yes. But enterprise domains don't have any additional config. So your options, from low to high effort, are:
1) Don't add a config at all
2) Add a boolean field `domain_uses_dialer` to `DomainPermissionsMirror` and make it visible in django admin
3) Add the boolean field and also add a checkbox to the `DomainPermissionsMirrorView`, which is currently a read-only UI.

I think any of these is fine. `DomainPermissionsMirror` is new enough and invisible enough that no one is using it at this point except the projects it was built for.

ctsims committed:
Fix removed formplayer header

ctsims committed:
Merge branch 'master' into ctsims/call_center_widgets

ctsims committed:
Merge branch 'master' into ctsims/call_center_widgets

ctsims committed:
explicit filetype indicator for static ref

ctsims committed:
Include shared tags module

orangejenny commented on corehq/apps/widget/forms.py:
```diff
@@ -0,0 +1,95 @@
+            hqcrispy.B3MultiField(
+                _("Telephony Services"),
+                hqcrispy.InlineField(
+                    'is_enabled', data_bind="checked: isEnabled"
```
Are these `data_bind` options intentional? It doesn't look like you're using knockout.

ctsims replied:
Just boilerplate from somewhere else (which also set KO bindings it didn't use). Will change.

orangejenny commented on corehq/apps/widget/forms.py:
```diff
@@ -0,0 +1,95 @@
+        ]
+
+    def __init__(self, data, *args, **kwargs):
+        self._domain = kwargs.pop('domain')
```
Nitpick: why `_domain` instead of `domain`? Most forms use the latter.

ctsims replied:
Just boilerplate from somewhere else. Will change.

orangejenny commented on corehq/apps/widget/static/widget/js/testUDPNetwork.js:
```diff
@@ -0,0 +1,209 @@
+/**
+Based on AWS boilerplate
```
Should this file be committed to production? Looks like it's only referenced in `checkNetwork` which is never called. This seems like it would be useful to have as documentation, but not especially useful checked in.

ctsims committed:
Test deploy

ctsims committed:
add back css compress

orangejenny reviewed (commented):
Our standard for new js files is that they're compatible with RequireJS. The two minified files and detectRTC.js look like they support AMD modules already, and then `ccputil.js` contains the entry point (`initDialer`), so you could make that your RequireJS main module. Overall, it'd look something like this:
```
hqDefine("widgets/js/ccputil", [
   "hqwebapp/js/initial_page_data",
    "widgets/js/detectRTC",
   "widgets/js/amazon-connect-min",
    "widgets/js/connect-streams-min",
], function (
   initialPageData,
   DetectRTC
   /* not sure if you need params for the other two files, depends on if the code references them modules directly */
) {
   /* current code, ideally excluding the dead functions like copyToClipboard */

    $(function () {
        initDialer(initialPageData.get('callout_number'), initialPageData.get('aws_instance_id'))
    });
});
```

And then instead of the inline js in `web_app_dialer.html`, you'd use `requirejs_main` and initial page data:
```
{# this goes near the top of the page, outside of any other block #}
{% requirejs_main "widgets/js/web_app_dialer" %}

{# these go inside widget_content or some other block #}
{% initial_page_data 'callout_number' callout_number %}
{% initial_page_data 'aws_instance_id' dialer_settings.aws_instance_id %}
```

ctsims committed:
removing compress due to deploy issues

ctsims committed:
Update JS dependency pattern to require.js

ctsims committed:
remove unused databinds

ctsims committed:
Update domain reference pattern

ctsims committed:
remove redundant custom logo context fetch

ctsims commented on corehq/apps/widget/views.py:
```diff
@@ -7,20 +7,29 @@
 from memoized import memoized
 
 
+#TODO: Add toggle decorator
```
The dialer doesn't actually load any meaningful context from web apps (just styles) so I think the current toggle set is fine. Will want to go more fine grained in the future if we actually allow the dialer to extend permissions to any AWS Connect stuff, (right now the user still needs their session credentials logged in anyway) or if we start allowing them to do things like backreference case ID's

ctsims committed:
remove unused templating

ctsims committed:
linter

ctsims committed:
Linting and mild refactoring

ctsims committed:
linting

ctsims committed:
linting

ctsims committed:
linting

ctsims committed:
Explicit global

ctsims committed:
Don't lint boilerplant / library files

ctsims committed:
Repackage under the integrations app

ctsims committed:
repackange under the integrations app

ctsims committed:
linting

ctsims committed:
JS Strictness

ctsims commented:
@orangejenny Did a pass based on initial feedback:

- JS Imports are now managed Require.js modules
- I moved everything out of the new `widget` module and into the existing `integration` module (previously just housing the SimPrints configuration models), which I think most cleanly captures the external dependencies 
- Cleaned up some of the template code I'd copied from other spaces
- I fixed all of the linter inputs for code that I actually wrote vs am copying-in from AWS. I'm still not particularly inclined to go scrub all of the boilerplate code much more, since I think it's highly diminishing returns to do that v. re-implement views, and I don't think either are worth the time on the first pass, since the code is fairly isolated. 

I've put everything on staging and confirmed that it all works as expected.

orangejenny commented on urls.py:
```diff
@@ -90,7 +90,7 @@
     url(r'^translations/', include('corehq.apps.translations.urls')),
     url(r'^submit_feedback/$', submit_feedback, name='submit_feedback'),
-    url(r'^widget/', include('corehq.apps.widget.urls')),
+    url(r'^integration/', include('corehq.apps.integration.urls')),
```
The integration urls are already included in `domain.urls`, which is correct for your new code since `DialerSettingsView` is domain-specific (I didn't notice on first review). I think you can just remove this and everything will continue to work, just with a different URL.

https://github.com/dimagi/commcare-hq/blob/47a7dd6948f49b0dd9e7a5aeca1b795b348cef48/corehq/apps/domain/urls.py#L182

orangejenny commented:
@ctsims Thank you for updating. I had one comment about the urls, but once that's fixed, this'll be good to go.

> I fixed all of the linter inputs for code that I actually wrote vs am copying-in from AWS. I'm still not particularly inclined to go scrub all of the boilerplate code much more, since I think it's highly diminishing returns to do that v. re-implement views, and I don't think either are worth the time on the first pass, since the code is fairly isolated.

This code definitely violates the principle of "A codebase should look like it was written by one person," and I don't love that, but I agree this is isolated and that there's little value to rewriting it. RequireJS was the major thing, since without those changes someone would need to migrate these files later. The remaining differences are stylistic (naming conventions, spacing) or are because the code doesn't use our common dependencies (jQuery, moment).

One remaining question: are translations relevant for this code? That would be a matter of reading through the js files and wrapping any user-facing text with a `gettext`.

ctsims committed:
Merge branch 'master' into ctsims/call_center_widgets

ctsims committed:
Move DetectRTC dependency to yarn management

ctsims committed:
Basic translation support

ctsims commented:
@orangejenny Translations submitted so I think this is in its final form

ctsims commented:
@orangejenny pending feedback from biyeun on that question anyhow.

orangejenny reviewed (approved):
Looks good. I think https://github.com/dimagi/commcare-hq/pull/28168 will resolve the yarn question.

ctsims committed:
Merge branch 'master' into ctsims/call_center_widgets

ctsims committed:
Merge branch 'master' into ctsims/call_center_widgets
